src/dawn/IIR/IIRNodeIterator.h

```diff
@@ -0,0 +1,157 @@
+    iterator_ = (root->childrenBegin());
+    end_ = (root->childrenEnd());
+    restIterator_ =
+        IIRNodeIterator<typename RootIIRNode::ChildType, LeafNode>(root->childrenBegin()->get());
+  }
+
+  IIRNodeIterator() = default;
+  IIRNodeIterator(IIRNodeIterator&&) = default;
+  IIRNodeIterator(const IIRNodeIterator&) = default;
+  IIRNodeIterator& operator=(IIRNodeIterator&&) = default;
+
+  IIRNodeIterator& operator++() {
+    ++restIterator_;
+    if(restIterator_.isEnd()) {
+      ++iterator_;
+
+      if(!isEnd()) {
+        restIterator_ =
+            IIRNodeIterator<typename RootIIRNode::ChildType, LeafNode>(iterator_->get());
+      }
+    }
+    return *this;
+  }
+
+  bool operator==(const IIRNodeIterator& other) const {
+    return (iterator_ == other.iterator_) && (restIterator_ == other.restIterator_);
+  }
+
+  bool operator!=(const IIRNodeIterator& other) const { return !(*this == other); }
+
+  void setToEnd() {
+    iterator_ = end_;
+    // the recursive rest of iterators is constructed with last node (i.e. end-1) and set to end
+    // recursively
+    restIterator_ =
+        IIRNodeIterator<typename RootIIRNode::ChildType, LeafNode>(std::prev(iterator_)->get());
+    restIterator_.setToEnd();
+  }
+
+  const std::unique_ptr<LeafNode>& operator*() {
+    return derefImpl<typename RootIIRNode::ChildType>();
+  }
+
+  template <typename T>
+  const std::unique_ptr<LeafNode>&
+  derefImpl(typename std::enable_if<std::is_same<T, LeafNode>::value>::type* = 0) {
+    return *iterator_;
+  }
+
+  template <typename T>
+  const std::unique_ptr<LeafNode>&
+  derefImpl(typename std::enable_if<!std::is_same<T, LeafNode>::value>::type* = 0) {
+    return *restIterator_;
+  }
+
+  bool isEnd() const { return iterator_ == end_; }
+};
+
+template <typename LeafNode>
+class IIRNodeIterator<LeafNode, LeafNode> {
+public:
+  IIRNodeIterator() = default;
+  IIRNodeIterator(const LeafNode* root) {}
+  IIRNodeIterator(IIRNodeIterator&&) = default;
+  IIRNodeIterator(const IIRNodeIterator&) = default;
+
+  IIRNodeIterator& operator=(IIRNodeIterator&&) = default;
+
+  IIRNodeIterator& operator++() { return *this; }
+
+  bool isEnd() const { return true; }
+  void setToEnd() {}
+  bool operator==(const IIRNodeIterator& other) const { return true; }
+  bool operator!=(const IIRNodeIterator& other) const { return true; }
+};
+
+template <typename RootNode, typename LeafNode>
+class IIRRange {
+  const RootNode& root_;
+
+public:
+  IIRRange(const RootNode& root) : root_(root) {}
+  IIRNodeIterator<RootNode, LeafNode> begin() {
+    return IIRNodeIterator<RootNode, LeafNode>(&root_);
+  }
+  IIRNodeIterator<RootNode, LeafNode> end() {
+    auto it = IIRNodeIterator<RootNode, LeafNode>(&root_);
+    it.setToEnd();
+    return it;
+  }
+};
+
+} // namespace iir
+
+template <typename Leaf, typename RootNode>
+iir::IIRRange<RootNode, Leaf> iterateIIROver(const RootNode& root) {
```

Comment:
are we sticking with this naming? My initial comment was:
 
I'm not sure about the name here, evtl use something that tells that you iterate over the children('s childern) of the current node? If I read `iteateIIROver<Stages>(*this)` this reads as iterating over all stages. I don't know the best solution yet